### PR TITLE
Fix handling of Control+C and Control+D

### PR DESCRIPTION
**Pull request by saagarjha**
This pull request fixes most of #212, as well as includes a fix for a typo that I noticed. I say "most" because <kbd>Control</kbd>+<kbd>C</kbd> doesn't interrupt the interpreter, as it does in other programs do. Is this something that's worth pursing?

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
> Is this something that's worth pursing?

Yes, definitely!

For ROOT we "simply" use good old `longjmp` from the signal handler. We can also throw a dedicated exception. Either way would work for me! But none of this has to happen in this PR.

File: lib/UserInterface/textinput/Text.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ namespace textinput {
 
     Text& operator+=(char C) { insert(length(), C); return *this; }
     Text& operator=(const std::string& S) {
-      // Assing string S to this, initialize with default colors.
```
Comment (Axel-Naumann) [inline lib/UserInterface/textinput/Text.h]: LOL thanks :-)

**Comment by Axel-Naumann** [comment]
We were missing a couple of commits in cling's textinput :-/ Could you rebase + fix the conflicts? Sorry about that!

**Comment by saagarjha** [comment]
I was reading your comments and GitHub suddenly started showing a merge conflict, so I was wondering what was going on there. I think I've fixed it–unfortunately I don't have resources to spare (hard drive space, mostly) to build the project with the conflict resolved, so I'm not sure that it compiles, but it looks about right…

I'd be fine with writing a signal handler for <kbd>Control</kbd>+<kbd>C</kbd>–but as I said, I can't build cling with debugging symbols and I won't be able to work on this for at least a couple days. It's your choice whether you want to merge this in now, or wait for me to finish the implementation for interrupting execution. By the way, if we _do_ write a signal handler for `SIGINT` there are a whole lot of restrictions on what it's allowed to do. If my memory serves right, C++ exceptions are not allowed, and `setjmp`/`longjmp` have a bunch of caveats. Do you have any thoughts on this, regarding what ROOT does, or should I go and look up the full list of restrictions that POSIX has and start from there?

**Comment by Axel-Naumann** [comment]
For ROOT we survived fine with `longjmp` even though it's evil and forbidden. Law vs practice, I suppose ;-)

I'm checking on https://github.com/root-project/root/pull/2142 how this PR does for ROOT.

**Comment by saagarjha** [comment]
So, should I start working on a solution that uses `setjmp`/`longjmp`, or what?

**Comment by Axel-Naumann** [comment]
Yes please! (But that's independent of this PR which I'll merge once Jenkins is convinced that it's all good for https://github.com/root-project/root/pull/2142 .)

**Comment by saagarjha** [comment]
What happened to all the new commits that I rebased in? They seem to have disappeared.

**Comment by Axel-Naumann** [comment]
Well - you're working off a "copy" of the master branch, bad things can happen ;-) I did not touch this PR, so what you see here is whatever Github made of your branch. But Github tells me about your master: "This branch is 7 commits ahead, 1 commit behind root-project:master." https://github.com/saagarjha/cling/commits/master only shows two commits by you. I.e. sorry, I don't think I can help with that, maybe you forgot to push?

**Comment by saagarjha** [comment]
No, you told me to rebase in some commits earlier (https://github.com/root-project/cling/pull/243/commits/7ceeb4b9fcaa39848a571c2f25b11a4fc92010ba, https://github.com/root-project/cling/pull/243/commits/89a82a257fe1fca7a15551b329022f9bfe2aab4d, https://github.com/root-project/cling/pull/243/commits/3b4ced228e900ae1c2c1b22aa5905109e855a4e4, https://github.com/root-project/cling/pull/243/commits/6c52d503fe56ecbad54fb4d4b666dd4dafc2c08b, and https://github.com/root-project/cling/pull/243/commits/4b3cabc98936861a992f6e7728aa6febf7ec2b60) to avoid a merge conflict with master on this repository, which I did. It seems like those commits disappeared recently (from this repository, not mine!). This makes it look like my pull request makes more changes than it should. I have not touched my repository since June 4th, so these changes must have been made after that.

**Comment by Axel-Naumann** [comment]
Merged by hand, no idea what happened here. Let's see whether the current state survives.

Thanks for your PR!

**Comment by saagarjha** [comment]
Thanks! I'm going to hard reset back to the current repository state and continue working on getting interrupting execution to work right.

**Comment by saagarjha** [comment]
Something odd sees to be going on here: I can no longer build cling on my machine. My guess is that the projects are out of sync in some way. Would you mind taking a look at this, @Axel-Naumann?

**Comment by Axel-Naumann** [comment]
- Can you send the commit hash for cling, clang, and llvm?
- How does it fail?
(It works for me, Jenkins and Travis.)

**Comment by saagarjha** [comment]
llvm is cling-patches@e0b472e46eb.
clang is master@58199aaa59.
cling is master@d57fbe37.

I'm running a single-threaded make, just so I can figure out what exactly is wrong, but last time I checked (with `make -j4`) it was something like:

```
2304:3: error: Couldn't find class 'ValuesCode'
  ValuesCode<[{
  ^
```

I'll update this once the build finishes.

**Comment by saagarjha** [comment]
Ok, so I'm getting the same error I was getting before in #212. It looks like someone has done a force push or similar to the llvm repository at http://root.cern.ch/git/llvm.git; the "latest" commit on master there is e0b472e46eb, but the remote points at 3c8af5480a2? Something like that. My git is very confused, because it looks like I pulled in changes that were expunged some time later, and it thinks that it's somehow even with but also behind the remote master branch.

**Comment by Axel-Naumann** [comment]
Weird; we didn't touch llvm since a long while because we use the "stable" 5.0 release. So a git pull should do for you?

**Comment by saagarjha** [comment]
I just nuked all the repositories and performed a clean build, and I was able to get it to work. But then I changed a couple lines, ran `make` again, and it started complaining about missing headers:

```
In file included from /Users/saagarjha/Git/llvm-cling/src/tools/cling/lib/Interpreter/AutoSynthesizer.cpp:10:
In file included from /Users/saagarjha/Git/llvm-cling/src/tools/cling/lib/Interpreter/AutoSynthesizer.h:13:
/Users/saagarjha/Git/llvm-cling/src/tools/cling/lib/Interpreter/ASTTransformer.h:15:10: fatal error:
      'clang/AST/Decl.h' file not found
#include "clang/AST/Decl.h" // for Result(Decl)
         ^~~~~~~~~~~~~~~~~~
```

Something's really finicky with the build system, because it's not reproducible on my machine at all–maybe I'm just using the wrong commits? I'm essentially following the build instructions [here](https://root.cern.ch/cling-build-instructions), here's my full compile script just to be complete:

```sh
#!/bin/sh

mkdir ~/Git/llvm-cling
cd ~/Git/llvm-cling
git clone http://root.cern.ch/git/llvm.git src
cd src
git checkout cling-patches
cd tools
git clone https://github.com/saagarjha/cling.git # My fork, which appears to be up-to-date with this repository
git clone http://root.cern.ch/git/clang.git
cd clang
git checkout cling-patches
mkdir /tmp/build
cd /tmp/build
cmake -DCMAKE_BUILD_TYPE=DEBUG ~/Git/llvm-cling/src/
make -j 4
```

I'm building with:
* llvm on cling-patches@e0b472e46eb
* clang on cling-patches@1b5a5b2fd1
* cling on master@891b279b

Could you please let me know if that's what you're building with? If not, could you send me the SHAs of the whatever you're on, as well as your CI build script? I really want to write this fix up but I can't do so if it doesn't compile (well, I could, but I'm sure you'd prefer if I submitted a patch that built and was tested 😉)

P.S. slightly unrelated, but clones from http://root.cern.ch are kinda slow (100 KB/s). Is this known behavior, and if so, is there a faster mirror?

**Comment by Axel-Naumann** [comment]
> 'clang/AST/Decl.h' file not found
means that we have a build system issue; libInterpreter thinks it doesn't depend on clangHeaders

> `cmake -DCMAKE_BUILD_TYPE=DEBUG ~/Git/llvm-cling/src/`
I always call that `-DCMAKE_BUILD_TYPE=Debug` (lower-case "ebug")?

The commit hashes are correct.

root.cern.ch is not meant to be slow at all :-) What's your traceroute? What's your throughput (say wget) for http://cern.ch/root-files/tp_process_imt_small.root ?

**Comment by saagarjha** [comment]
> The commit hashes are correct.

Hmm, what build command are you using? The odd thing is that I can get this to work sometimes, but other times it just fails to compile.

> root.cern.ch is not meant to be slow at all :-)

I'm on a university network, and it's very late at night, so being the only one on the network I'm able to saturate my gigabit ethernet connection pretty easily with any other website. However, my connection with CERN seems to be capped at 100 KB/s for some reason, no matter what I download. I can go talk to my network administrator later today to see if the throttling is done by them, if you're sure that it's not your end that's rate limiting me. The traceroute, by the way, seems to hit CERN's network very quickly though the Energy Science Network, but then it gets stuck somewhere inside there:

```
$ traceroute root.cern.ch
traceroute to p05153026283679.cern.ch (137.138.13.143), 64 hops max, 52 byte packets
 1  resnet-8-1.resnet.ucsb.edu (169.231.8.1)  1.162 ms  0.953 ms  0.550 ms
 2  128.111.174.1 (128.111.174.1)  1.427 ms  1.623 ms  1.492 ms
 3  lax-hpr3--ucsb-100ge.cenic.net.cenic.net (137.164.26.237)  3.800 ms  3.722 ms  3.523 ms
 4  hpr-svl-hpr3--lax-hpr3-100ge.cenic.net (137.164.25.74)  11.296 ms  11.637 ms  11.384 ms
 5  hpr-esnet--svl-hpr2-100ge.cenic.net (137.164.26.10)  12.606 ms  11.884 ms  11.696 ms
 6  sacrcr5-ip-a-sunncr5.es.net (134.55.40.5)  14.518 ms  15.101 ms  14.639 ms
 7  denvcr5-ip-a-sacrcr5.es.net (134.55.50.202)  35.407 ms  35.245 ms  35.454 ms
 8  kanscr5-ip-a-denvcr5.es.net (134.55.49.58)  45.752 ms  45.961 ms  45.801 ms
 9  chiccr5-ip-a-kanscr5.es.net (134.55.43.81)  57.441 ms  58.272 ms  57.169 ms
10  eqxchicr5-ip-b-chiccr5.es.net (134.55.218.62)  57.219 ms
    washcr5-ip-a-chiccr5.es.net (134.55.36.46)  74.015 ms
    eqxchicr5-ip-b-chiccr5.es.net (134.55.218.62)  57.194 ms
11  * * eqxashcr5-ip-a-eqxchicr5.es.net (134.55.207.53)  73.644 ms
12  washcr5-ip-c-eqxashcr5.es.net (134.55.219.1)  73.970 ms
    londcr5-ip-c-aofacr5.es.net (134.55.37.22)  148.587 ms  148.496 ms
13  * * cern773cr5-ip-a-londcr5.es.net (134.55.42.58)  163.076 ms
14  e773-e-rbrxl-2-te5.cern.ch (192.65.184.193)  161.993 ms
    londcr5-ip-b-aofacr5.es.net (134.55.39.181)  148.167 ms  148.076 ms
15  e513-e-rbrxl-2-ne7.cern.ch (192.65.184.174)  162.594 ms  162.457 ms
    cern773cr5-ip-a-londcr5.es.net (134.55.42.58)  162.327 ms
16  * e773-e-rbrxl-2-te5.cern.ch (192.65.184.193)  161.693 ms  162.184 ms
17  * * e513-e-rbrxl-2-ne7.cern.ch (192.65.184.174)  161.608 ms
18  * * *
19  * * *
20  * * *
21  * * *
22  * * *
23  * * *
24  * * *
25  * * *
26  * * *
27  * * *
28  * * *
29  *^C
```

I don't want to waste your time solving network issues, though, so if there's someone else responsible for I can try talking to them instead.

**Comment by Axel-Naumann** [comment]
Right, we're hiding. But we have a pretty good connection (and that's an understatement) with ESnet. If you see the same terrible throughput from cern.ch as on root.cern.ch then it's very likely between you and ESnet. Try https on https://cern.ch/root-files/tp_process_imt_small.root - if that accelerates things then it's a deep package inspection on your side...

**Comment by saagarjha** [comment]
Unfortunately, I'm still seeing slow speeds under HTTPS as well. I talked to my network administrator and they said it must be something on CERN's side, so I think I'll send their service desk an email.

**Comment by Axel-Naumann** [comment]
Look - I wasn't kidding when I said that ESnet to CERN is pretty good. This is part of the LHC data highway: https://my.es.net/ No way is there a problem or all of LHC would notice. Good luck hunting down the bottleneck!

**Comment by saagarjha** [comment]
> means that we have a build system issue; libInterpreter thinks it doesn't depend on clangHeaders

Anything I can do about this? This is happening consistently now. I'm really sorry to keep bugging you with this issue, but I just can't seem to get around it at all no matter how many times I reclone and rebuild.

> Look - I wasn't kidding when I said that ESnet to CERN is pretty good.

I'm not doubting you on that–with the amount of data you handle I'm sure you have the biggest, baddest connection possible. I've sent an email to CERN's service desk just to confirm they're not throttling me.

**Pull request closed without merging**